command.go

```diff
@@ -877,6 +877,13 @@ func (c *Command) ValidateArgs(args []string) error {
 	if c.Args == nil {
 		return nil
 	}
+	if len(c.ValidArgs) > 0 {
```

Comment:
What do you think about putting this logic into an un-exported function?  You could then invoke it like:
```
if err := validArgs(c, args); err != nil {
  return err
}
return c.Args(c, args)
```

Reply:
I think it is not required for now. Because this block is used in a single place in the code base, adding a branch through a call to a function adds unneccesary complexity to the code. It is likely to be optimized by the compiler, so it should have little to no impact on performance. But I would anyway wait until it is really required. Nonetheless, a comment can be added just before line 880, to let readers know that the block can be considered an atomic/independent operation.

Reply:
@jharshman, after your comment, I think that it can be improved. ATM, two for loops are used. One in `command.go` and another one in `cobra.go` (`func stringInSlice`). `stringInSlice` is only used twice: in `command.go` and `command_test.go`. So, we can merge `stringInSlice` in `command_test.go`, and replace the block you commented about with the following:

``` golang
package main

import (
	"fmt"
	"strings"
)

func main() {
	fmt.Println("Hello, playground")
	
	ValidArgs := []string{"one", "two", "three"}
	Args := []string{"three", "one", "a"}
	
    e := sliceContains(ValidArgs, Args)
    if e != nil {
	  fmt.Println(*e)
	}
}

func sliceContains(s, x []string) *string {
  const p := "$"
  for _, a := range x {
    if !strings.Contains(p+strings.Join(s, p)+p, p+a+p) {
      return &a
    }
  }
  return nil
}
```

Reply:
Looks good but what's the purpose of ` (p+strings.Join(s, p)+p, p+a+p)`? 
Why not `!strings.Contains(strings.Join(s,""), a)`?

Reply:
We need to ensure that the argument matches one of the strings in `ValidArgs` exactly:

``` go
package main

import (
	"fmt"
	"strings"
)

func main() {
	s := []string{"one", "two", "three"}
	a := "re"
	p := "$"	
	fmt.Println(strings.Contains(p+strings.Join(s, p)+p, p+a+p))
	fmt.Println(strings.Contains(strings.Join(s,""), a))
}
```

``` bash
false
true
```

Reply:
I see, you are delimiting with a dollar sign.

Reply:
That's it. Please, let me know if you find any other issue with this PR. Otherwise, I'd be really glad if you could approve it.

Reply:
👍

Reply:
I have to admit, I don't particularly like this way to search for valid args. First, we are still going to be doing a second loop, only now it's hidden inside Contains() https://golang.org/src/strings/strings.go?#L947 and second the code is less obvious. I STRONGLY prefer obvious to smart. Someone else is going to have to maintain this and simple is always going to be better. Third, I don't see any testing related to the usage of the `$` to prove it is correct. And the last thing that bothers me is that I think it's wrong. What if my validArgs = []string{"aa$bb"}. Wouldn't this code allow both "aa" and "bb" to pass, even though they are not in the ValidArgs?

I think the helper function was correct and clear and would prefer to continue something more obvious.

Reply:
It's ok. The change was in a separate commit, so it is dropped now. 

> Third, I don't see any testing related to the usage of the `$` to prove it is correct. 

There is quite a lot of testing, since may of the tests in `args_test.go`. Nonetheless, you are correct, corner cases such as a ValidArg containing `$` were not properly handled.

Please, let me know if you find any other issue with this PR.

Reply:
I'm all for a helper function here.